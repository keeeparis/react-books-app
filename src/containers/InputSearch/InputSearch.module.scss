.input {
  border: none;
  appearance: none;
  outline: none;
  font-family: inherit;
  font-size: 18px;
  padding: 10px 35px 10px 40px;
  position: relative;
  border-radius: 8px;
  box-shadow: 1px 6px 20px 1px rgb(0 0 0 / 14%);

  &:focus-within,
  &:focus-visible {
    border: none;
    outline: none;
  }
}

@media (max-width: 660px) {
  .input {
    padding: 15px 35px 15px 40px;
  }
}

.inputWrapper {
  position: relative;
  &::after {
    content: '';
    z-index: 1;
    position: absolute;
    top: 13px;
    left: 10px;
    width: 10px;
    height: 10px;
    border-bottom: 2px solid black;
    transform: rotate(-45deg);
  }

  &::before {
    content: '';
    z-index: 1;
    position: absolute;
    top: 17px;
    left: 17px;
    width: 6px;
    height: 6px;
    border-top: 2px solid black;
    border-right: 2px solid black;
    // transform: rotate(90deg);
  }
}

@media (max-width: 660px) {
  .inputWrapper {
    &::after {
      top: 19px;
    }
    &::before {
      top: 23px;
    }
  }
}

.Close {
  position: absolute;
  right: 12px;
  top: 30%;
  width: 16px;
  line-height: unset;
  appearance: none;
  height: 16px;
  transform: scale(1.3);
  border-radius: 50%;
  border: 1px solid black;
  color: black;
  outline: none;
  background-color: transparent;
  cursor: pointer;
  transition: all 0.2s ease;
  opacity: 0;
  visibility: hidden;

  &:hover,
  &:focus-within {
    background-color: var(--color-3);
    color: white;
    border-color: white;
  }

  &.visible {
    opacity: 1;
    visibility: visible;
  }
}

@media (max-width: 660px) {
  .Close {
    top: 35%;
  }
}
