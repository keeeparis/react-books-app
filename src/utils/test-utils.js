import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import 'regenerator-runtime/runtime'
import { Provider } from 'react-redux'
import { store } from '../redux/store/store'
import { MemoryRouter } from 'react-router-dom'
import '../i18n/config'

const AllTheProviders = ({ children }) => {
  return (
    <Provider store={store}>
      <MemoryRouter>{children}</MemoryRouter>
    </Provider>
  )
}

const customRender = (ui, options) => {
  return render(ui, { wrapper: AllTheProviders, ...options })
}

export * from '@testing-library/react'
export * from '@testing-library/jest-dom'
export { userEvent }
export * from 'regenerator-runtime/runtime'
export * from '../i18n/config'

export { customRender as render }
