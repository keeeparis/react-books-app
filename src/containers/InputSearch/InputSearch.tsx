import React, {
  InputHTMLAttributes,
  useState,
  MouseEvent,
  forwardRef,
} from 'react'
import cn from 'classnames'

import SearchHistory from '../../components/SearchHistory'
import useUpdateLocalStorage from '../../hooks/useUpdateLocalStorage'
import styles from './InputSearch.module.scss'

import { removeSearchHistoryFromLocalstorage } from '../../utils/searchHistoryLocalStorage'

interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  resetInput: () => void
  stateInput: string
  handleHistorySearchClick: (option: string) => () => void
}

const InputSearch = forwardRef<HTMLInputElement, InputProps>(
  (
    { value, resetInput, stateInput, handleHistorySearchClick, ...props },
    ref
  ) => {
    const [isHistoryVisible, setIsHistoryVisible] = useState(false)
    const [history, setHistory] = useState<string[] | null>(null)

    /* Когда отправляется новый запрос, мы меняем инпут в нашем store,
    и триггерится данный хук, который обновляет данные с localstorage 
    в state history */
    useUpdateLocalStorage('search-history', setHistory, stateInput)

    const deleteFromSearchHistory =
      (option: string) => (e: MouseEvent<HTMLButtonElement>) => {
        e.stopPropagation()
        removeSearchHistoryFromLocalstorage(option)
        setHistory(JSON.parse(localStorage.getItem('search-history')!))
      }

    return (
      <div className={styles.inputWrapper}>
        <input
          value={value}
          className={styles.input}
          role="input"
          onFocus={() => setIsHistoryVisible(true)}
          onBlur={() => setIsHistoryVisible(false)}
          ref={ref}
          {...props}
        />

        {/* Reset Button Within Input */}
        <button
          className={cn(styles.Close, { [styles.visible]: !!value })}
          onClick={resetInput}
          type="reset"
        >
          &#215;
        </button>

        {/* Search History Results */}
        {history && (
          <SearchHistory
            history={history}
            historyVisible={isHistoryVisible}
            deleteFromSearchHistory={deleteFromSearchHistory}
            handleHistorySearchClick={handleHistorySearchClick}
          />
        )}
      </div>
    )
  }
)

export default InputSearch
