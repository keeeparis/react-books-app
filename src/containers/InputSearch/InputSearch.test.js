import React from 'react'
import { render, screen, fireEvent, userEvent } from 'test-utils'

import InputSearch from './InputSearch'

const props = {
  value: '',
  onChange: jest.fn(),
}

beforeEach(() => {})

describe('InputSearch', () => {
  test('calls the onChange callback handler', async () => {
    const { asFragment } = render(<InputSearch {...props} />)
    await userEvent.type(screen.getByRole('input'), 'Steve')
    expect(props.onChange).toBeCalledTimes(5)
    expect(asFragment()).toMatchSnapshot()
  })
})
