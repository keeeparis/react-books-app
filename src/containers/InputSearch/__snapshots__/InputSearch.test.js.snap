// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputSearch calls the onChange callback handler 1`] = `
<DocumentFragment>
  <div
    class="inputWrapper"
  >
    <input
      class="input"
      role="input"
      value=""
    />
    <button
      class="Close"
      role="resetButton"
      type="reset"
    >
      Ã—
    </button>
  </div>
</DocumentFragment>
`;
